libnmstate/nm/connection.py

```diff
@@ -21,157 +21,171 @@
 from . import nmclient
 
 
-def create_profile(settings):
-    con_profile = nmclient.NM.SimpleConnection.new()
-    for setting in settings:
-        con_profile.add_setting(setting)
-
-    return con_profile
-
-
-def add_profile(connection_profile, save_to_disk=True):
-    client = nmclient.client()
-    mainloop = nmclient.mainloop()
-    user_data = mainloop
-    mainloop.push_action(
-        client.add_connection_async,
-        connection_profile,
-        save_to_disk,
-        mainloop.cancellable,
-        _add_connection_callback,
-        user_data,
-    )
-
-
-def _add_connection_callback(src_object, result, user_data):
-    mainloop = user_data
-    try:
-        con = src_object.add_connection_finish(result)
-    except Exception as e:
-        if mainloop.is_action_canceled(e):
-            logging.debug(
-                'Connection adding canceled: error=%s', e)
+class ConnectionProfile(object):
+
+    def __init__(self, profile=None):
+        self._con_profile = profile
+        self._nmclient = nmclient.client()
+        self._mainloop = nmclient.mainloop()
+
+    def create(self, settings):
+        self._con_profile = nmclient.NM.SimpleConnection.new()
+        for setting in settings:
+            self._con_profile.add_setting(setting)
+
+    def import_by_device(self, nmdev):
+        self._con_profile = None
+        ac = get_device_active_connection(nmdev)
+        if ac:
+            self._con_profile = ac.props.connection
+
+    def import_by_id(self, con_id):
+        self._con_profile = None
+        if con_id:
+            self._con_profile = self._nmclient.get_connection_by_id(con_id)
+
+    def update(self, con_profile):
+        self._con_profile.replace_settings_from_connection(con_profile.profile)
+
+    def add(self, save_to_disk=True):
+        user_data = self._mainloop
+        self._mainloop.push_action(
+            self._nmclient.add_connection_async,
+            self._con_profile,
+            save_to_disk,
+            self._mainloop.cancellable,
+            self._add_connection_callback,
+            user_data,
+        )
+
+    def commit(self, save_to_disk=True, nmdev=None):
+        user_data = self._mainloop, nmdev
+        self._mainloop.push_action(
+            self._con_profile.commit_changes_async,
+            save_to_disk,
+            self._mainloop.cancellable,
+            self._commit_changes_callback,
+            user_data,
+        )
+
+    @property
+    def profile(self):
```

Comment:
Why make the profile first internal and then provide it to the outside nonetheless?

Reply:
Because it is part of the object state and it is currently needed by others outside the object.
As an example, the profile is loaded as part of the object initialization (import or create).